libtextsecure/message_receiver.js

```diff
@@ -309,8 +360,11 @@ MessageReceiver.prototype.extend({
             var ev = new Event('error');
             ev.error = error;
             ev.proto = envelope;
-            this.dispatchEvent(ev);
-            return Promise.reject(error);
+
+            var returnError = function() {
+                return Promise.reject(error);
+            };
+            this.dispatchAndWait(ev).then(returnError, returnError);
```

Comment:
I think you want a return here. Otherwise the call to decrypt() (e.g., in handleLegacyMessage) will resolve to undefined plaintext if the decrypt failed.